File: src/test/com/intellij/lang/jsgraphql/resolve/GraphQLSchemaResolveTest.java
package com.intellij.lang.jsgraphql.resolve;

import com.intellij.lang.jsgraphql.GraphQLResolveTestCaseBase;
import com.intellij.lang.jsgraphql.psi.*;

public class GraphQLSchemaResolveTest extends GraphQLResolveTestCaseBase {
    @Override
    protected String getBasePath() {
        return "/resolve/schema";
    }

    public void testGithubSchema() {
        doHighlightingTest();
    }

    public void testAniListSchema() {
        doHighlightingTest();
    }

    public void testBitQuerySchema() {
        doHighlightingTest();
    }

    public void testUniverseSchema() {
        doHighlightingTest();
    }

    public void testDirectiveArgumentObjectFieldValue() {
        doResolveWithOffsetTest(GraphQLEnumValueDefinition.class, "owner");
    }

    public void testDirectiveArgumentObjectField() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "allow");
    }

    public void testDefaultArgumentObjectField() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "field");
    }

    public void testDirectiveArgumentEnumValue() {
        doResolveWithOffsetTest(GraphQLEnumValueDefinition.class, "FOREVER");
    }

    public void testDefaultArgumentObjectFieldInsideArray() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "field");
    }

    public void testDefaultArgumentObjectFieldNestedCircular() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "order");
    }

    public void testObjectFieldDefinitionObjectType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "Location");
    }

    public void testObjectFieldDefinitionEnumType() {
        doResolveWithOffsetTest(GraphQLEnumTypeDefinition.class, "CheckStatusState");
    }

    public void testObjectFieldDefinitionScalarType() {
        doResolveWithOffsetTest(GraphQLScalarTypeDefinition.class, "Upload");
    }

    public void testObjectFieldDefinitionUnionType() {
        doResolveWithOffsetTest(GraphQLUnionTypeDefinition.class, "Node");
    }

    public void testObjectFieldDefinitionInterfaceType() {
        doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition.class, "Node");
    }

    public void testArgumentDefinitionObjectType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "Argument");
    }

    public void testArgumentDefinitionInputType() {
        doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition.class, "Argument");
    }

    public void testDirectiveArgument() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "fieldType");
    }

    public void testSchemaOperationQueryType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomQuery");
    }

    public void testSchemaOperationMutationType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomMutation");
    }

    public void testSchemaOperationSubscriptionType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomSubscription");
    }

    public void testSchemaExtensionOperationQueryType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomQuery");
    }

    public void testSchemaExtensionOperationMutationType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomMutation");
    }

    public void testSchemaExtensionOperationSubscriptionType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "CustomSubscription");
    }

    public void testInterfaceFieldDefinitionObjectType() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "Location");
    }

    public void testInputFieldDefinitionInputType() {
        doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition.class, "LocationInput");
    }

    public void testObjectTypeImplementsInterface() {
        doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition.class, "Entity");
    }

    public void testObjectTypeDirective() {
        doResolveWithOffsetTest(GraphQLDirectiveDefinition.class, "table");
    }

    public void testObjectFieldDefinitionDirective() {
        doResolveWithOffsetTest(GraphQLDirectiveDefinition.class, "field");
    }

    public void testInterfaceTypeImplementsInterface() {
        doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition.class, "Node");
    }

    public void testInterfaceTypeImplementsCircular() {
        doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition.class, "Named");
    }

    public void testInputObjectTypeCircular() {
        doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition.class, "Example");
    }

    public void testInputObjectTypeCircularList() {
        doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition.class, "Example");
    }

    public void testUnionTypeMember() {
        doResolveWithOffsetTest(GraphQLObjectTypeDefinition.class, "B");
    }

    public void testUnionTypeMemberCircular() {
        doResolveWithOffsetTest(GraphQLUnionTypeDefinition.class, "U");
    }
}

File: tests/testSrc/com/intellij/lang/jsgraphql/resolve/GraphQLSchemaResolveTest.kt
package com.intellij.lang.jsgraphql.resolve

import com.intellij.lang.jsgraphql.GraphQLResolveTestCaseBase
import com.intellij.lang.jsgraphql.psi.GraphQLDirectiveDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLEnumTypeDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLEnumValueDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLInputObjectTypeDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLInputValueDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLInterfaceTypeDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLObjectTypeDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLScalarTypeDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLUnionTypeDefinition
import com.intellij.openapi.progress.runBlockingCancellable

class GraphQLSchemaResolveTest : GraphQLResolveTestCaseBase() {
  override fun getBasePath(): String {
    return "/resolve/schema"
  }

  fun testGithubSchema() = runBlockingCancellable {
    doHighlightingTest()
  }

  fun testAniListSchema() = runBlockingCancellable {
    doHighlightingTest()
  }

  fun testBitQuerySchema() = runBlockingCancellable {
    doHighlightingTest()
  }

  fun testUniverseSchema() = runBlockingCancellable {
    doHighlightingTest()
  }

  fun testDirectiveArgumentObjectFieldValue() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLEnumValueDefinition::class.java, "owner")
  }

  fun testDirectiveArgumentObjectField() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "allow")
  }

  fun testDefaultArgumentObjectField() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "field")
  }

  fun testDirectiveArgumentEnumValue() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLEnumValueDefinition::class.java, "FOREVER")
  }

  fun testDefaultArgumentObjectFieldInsideArray() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "field")
  }

  fun testDefaultArgumentObjectFieldNestedCircular() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "order")
  }

  fun testObjectFieldDefinitionObjectType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "Location")
  }

  fun testObjectFieldDefinitionEnumType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLEnumTypeDefinition::class.java, "CheckStatusState")
  }

  fun testObjectFieldDefinitionScalarType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLScalarTypeDefinition::class.java, "Upload")
  }

  fun testObjectFieldDefinitionUnionType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLUnionTypeDefinition::class.java, "Node")
  }

  fun testObjectFieldDefinitionInterfaceType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition::class.java, "Node")
  }

  fun testArgumentDefinitionObjectType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "Argument")
  }

  fun testArgumentDefinitionInputType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition::class.java, "Argument")
  }

  fun testDirectiveArgument() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "fieldType")
  }

  fun testSchemaOperationQueryType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomQuery")
  }

  fun testSchemaOperationMutationType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomMutation")
  }

  fun testSchemaOperationSubscriptionType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomSubscription")
  }

  fun testSchemaExtensionOperationQueryType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomQuery")
  }

  fun testSchemaExtensionOperationMutationType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomMutation")
  }

  fun testSchemaExtensionOperationSubscriptionType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "CustomSubscription")
  }

  fun testInterfaceFieldDefinitionObjectType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "Location")
  }

  fun testInputFieldDefinitionInputType() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition::class.java, "LocationInput")
  }

  fun testObjectTypeImplementsInterface() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition::class.java, "Entity")
  }

  fun testObjectTypeDirective() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLDirectiveDefinition::class.java, "table")
  }

  fun testObjectFieldDefinitionDirective() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLDirectiveDefinition::class.java, "field")
  }

  fun testInterfaceTypeImplementsInterface() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition::class.java, "Node")
  }

  fun testInterfaceTypeImplementsCircular() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInterfaceTypeDefinition::class.java, "Named")
  }

  fun testInputObjectTypeCircular() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition::class.java, "Example")
  }

  fun testInputObjectTypeCircularList() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLInputObjectTypeDefinition::class.java, "Example")
  }

  fun testUnionTypeMember() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLObjectTypeDefinition::class.java, "B")
  }

  fun testUnionTypeMemberCircular() = runBlockingCancellable<Unit> {
    doResolveWithOffsetTest(GraphQLUnionTypeDefinition::class.java, "U")
  }
}
